// oicq-guild/lib/message.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GuildMessage = void 0;
const oicq_1 = require("oicq");
const message_1 = require("oicq/lib/message");
const common_1 = require("oicq/lib/common");
class GuildMessage {
    constructor(proto) {
        const head1 = proto[1][1][1];
        const head2 = proto[1][1][2];
        if (head2[1] !== 3840)
            throw new Error("unsupport guild message type");
        const body = proto[1][3];
        const extra = proto[1][4];
        this.guild_id = String(head1[1]);
        this.channel_id = String(head1[2]);
        this.guild_name = String(extra[2]);
        this.channel_name = String(extra[3]);
        this.sender = {
            tiny_id: String(head1[4]),
            nickname: String(extra[1])
        };
        this.seq = head2[4];
        this.rand = head2[3];
        this.time = head2[6];
        const parsed = (0, message_1.parse)(body[1]);
        this.message = parsed.message;
        this.raw_message = parsed.brief;
        (0, common_1.lock)(this, "proto");
    }
    /** 暂时仅支持发送： 文本、AT、表情 */
    async reply(content) {
        throw new oicq_1.ApiRejection(-999999, "no channel");
    }
}
exports.GuildMessage = GuildMessage;
